src/main/java/uk/gov/cslearning/catalogue/api/CourseController.java

```diff
@@ -245,15 +245,31 @@ public CourseController(CourseRepository courseRepository, CourseService courseS
     public ResponseEntity<PageResults<Course>> listMandatory(@RequestParam("department") String department,
             @RequestParam(value = "status", defaultValue = "Published") String status,
             Pageable pageable) {
-        LOGGER.debug("Listing mandatory courses for department {}", department);
         List<String> organisationParents = courseService.getOrganisationParents(department);
-
-        List<Course> courses = new ArrayList<>();
-        for (String parent : organisationParents) {
-            courses.addAll(courseService.fetchMandatoryCourses(status, parent));
-        }
-
-        return ResponseEntity.ok(new PageResults<>(courseService.prepareCoursePage(pageable, courses), pageable));
+        LOGGER.debug("Listing mandatory courses for department {} and its parent organisations {}", department, organisationParents);
+        List<Course> courses = courseRepository.findMandatoryOfMultipleDepts(organisationParents, "Published", PageRequest.of(0, 10000));
+        Map<String, Audience> courseAudiences = new HashMap<>();
+        courses.forEach(course -> courseService.getRelevantAudienceForCourse(course, organisationParents, courseAudiences));
```

Comment:
Could we make getRelevantAudienceForCourse return just the audience it finds? And then do the .put(course.getId(), audience) in the forEach lambda? I think passing the HashMap into the function makes it a bit unclear what it does (I know it was like this before but it'd be good to move away from it) - also makes it a little bit more tricky to unit test.

Reply:
Review comment is applied.